In [1]:
import pandas as pd 
import numpy as np 
from matplotlib import pyplot as plt
from sklearn import metrics as m 

In [2]:
d1=pd.read_csv('Advertising_Sales-4.csv')

In [3]:
d1.corr()

Unnamed: 0        TV     Radio  Newspaper     Sales
Unnamed: 0    1.000000  0.017715 -0.110680  -0.154944 -0.051616
TV            0.017715  1.000000  0.054809   0.056648  0.782224
Radio        -0.110680  0.054809  1.000000   0.354104  0.576223
Newspaper    -0.154944  0.056648  0.354104   1.000000  0.228299
Sales        -0.051616  0.782224  0.576223   0.228299  1.000000

In [4]:
x=d1[["TV"]]

In [5]:
y=d1.Sales

In [6]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=1 )

In [7]:
from sklearn.linear_model import LinearRegression
linreg=LinearRegression()
linreg.fit(x_train,y_train)

LinearRegression()

In [8]:
y_pred=linreg.predict(x_test)

In [9]:
from sklearn import metrics as m 

In [10]:
m.mean_squared_error(y_test,y_pred)

10.859345325608269

In [11]:
np.sqrt(m.mean_squared_error(y_test,y_pred ))

3.2953520791575928

x_train, y_train --->model building

x_test, y_test(actual values)-----> y_pred=predict(x_test)--->y_pred(predicted values)
r2_score(y_test,y_pred)---->accuracy on testing dataset i.e. x_test

x_train, y_train(actual values)-----> y_pred_train=predict(x_train)-----> y_pred_train(predicted values)
r2_score(y_train,y_pred_train)------>accuracy on training dataset i.e. x_train

In [12]:
m.r2_score(y_test,y_pred)  # accuracy on testing dataset

0.41535307148347866

In [13]:
y_train_pred=linreg.predict(x_train)

In [14]:
np.sqrt(m.mean_squared_error(y_train,y_train_pred ))

3.2333166589697684

In [15]:
m.r2_score(y_train,y_train_pred)  #accuracy on training dataset

0.6393874827744095

# Multiple Linear Regression

In [16]:
x=d1[['TV','Radio','Newspaper']]

In [17]:
y=d1.Sales

In [18]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=1 )

In [19]:
linreg=LinearRegression()
linreg.fit(x_train,y_train)

LinearRegression()

In [20]:
print(linreg.intercept_)  #b

2.9079470208164295


In [21]:
print(linreg.coef_) #a1- coeff of TV,a2- coeff of Radio,a3- coeff of Newspaper

[0.0468431  0.17854434 0.00258619]


Sales=0.0468431 X TV + 0.17854434  X Radio + 0.00258619 X Newspaper + 2.9079470208164295

In [22]:
y_pred=linreg.predict(x_test) #x_test-->TV, Radio and Newspaper- 40 records ,(actual values) y_test-->sales

In [23]:
m.mean_squared_error(y_test,y_pred)

1.9918855518287881

In [24]:
np.sqrt(m.mean_squared_error(y_test,y_pred ))

1.4113417558581578

Total records= 200 records
Dataset 1: Training Dataset (160 records) (Model has seen this data in training phase)
Dataset 2: Testing Dataset  (40 records)  (Unseen data--> Model has no knowledge of this data)

In [25]:
y_train_pred= linreg.predict(x_train) # training dataset(x_train and y_train)
print(np.sqrt(m.mean_squared_error(y_train,y_train_pred)))
print(m.r2_score(y_train, y_train_pred)) # accuracy on training dataset

y_test_pred= linreg.predict(x_test)
print(np.sqrt(m.mean_squared_error(y_test,y_test_pred))) 
print(m.r2_score(y_test, y_test_pred))# accuracy on testing dataset

#Variance=1.7369025901470923(bias of training) - 1.4113417558581578(Bias of testing)

1.7369025901470923
0.8959372632325174
1.4113417558581578
0.8927605914615385


In [26]:
print(linreg.intercept_)

2.9079470208164295


In [27]:
print(linreg.coef_)

[0.0468431  0.17854434 0.00258619]


In [28]:
m.r2_score(y_test,y_pred)

0.8927605914615385

# Ridge Regression

In [29]:
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
from sklearn.metrics import r2_score

rr = Ridge(alpha=0.01)
rr.fit(x_train, y_train) #RIDGE REGRESSION

pred_train_rr= rr.predict(x_train) #PREDICTIONS ON TRAINING SET

print('MSE: TRAINING SET')
print(np.sqrt(m.mean_squared_error(y_train,pred_train_rr)))
print('SCORE: TRAINING SET')
print(r2_score(y_train, pred_train_rr))

pred_test_rr= rr.predict(x_test) # PREDICTIONS ON TESTING TEST

print('MSE: TESTING SET')
print(np.sqrt(m.mean_squared_error(y_test,pred_test_rr))) 
print('SCORE: TESTING SET')
print(r2_score(y_test, pred_test_rr))

MSE: TRAINING SET
1.7369025901472777
SCORE: TRAINING SET
0.8959372632324952
MSE: TESTING SET
1.4113421493908445
SCORE: TESTING SET
0.8927605316571456


In [30]:
print(rr.intercept_)

2.907947854165572


In [31]:
print(rr.coef_)

[0.0468431  0.17854429 0.0025862 ]


# LASSO Regression

In [32]:
model_lasso = Lasso(alpha=0.01)
model_lasso.fit(x_train, y_train) #LASSO REGRESSION


pred_train_lasso= model_lasso.predict(x_train)   #PREDICTIONS ON TRAINING SET
print('MSE: TRAINING SET')
print(np.sqrt(m.mean_squared_error(y_train,pred_train_lasso)))
print('SCORE: TRAINING SET')
print(r2_score(y_train, pred_train_lasso))

pred_test_lasso= model_lasso.predict(x_test)  # PREDICTIONS ON TESTING TEST
print('MSE: TESTING SET')
print(np.sqrt(m.mean_squared_error(y_test,pred_test_lasso))) 
print('SCORE: TESTING SET')
print(r2_score(y_test, pred_test_lasso))

MSE: TRAINING SET
1.736902728711357
SCORE: TRAINING SET
0.8959372466289643
MSE: TESTING SET
1.4115135766636477
SCORE: TESTING SET
0.8927344786039069


In [33]:
print(model_lasso.intercept_)

2.90922799386513


In [34]:
print(model_lasso.coef_)

[0.04684247 0.17850829 0.00257448]


Multiple Linear Rgression- 0.0468431  0.17854434 0.00258619
LASSO Regression- 0.04684247 0.17850829 0.00257448
Ridge Regression-0.0468431  0.17854429 0.0025862

# LassoCV

In [35]:
from sklearn.linear_model import LassoCV
lasso_cv = LassoCV(alphas=[0.01, 0.1, 1.0, 10.0])

lasso_cv.fit(x_train, y_train) #RIDGE REGRESSION

pred_train_lasso_cv= lasso_cv.predict(x_train) #PREDICTIONS ON TRAINING SET

print('MSE: TRAINING SET')
print(np.sqrt(m.mean_squared_error(y_train,pred_train_lasso_cv)))
print('SCORE: TRAINING SET')
print(r2_score(y_train, pred_train_lasso_cv))

pred_test_lasso_cv= lasso_cv.predict(x_test) # PREDICTIONS ON TESTING TEST

print('MSE: TESTING SET')
print(np.sqrt(m.mean_squared_error(y_test,pred_test_lasso_cv))) 
print('SCORE: TESTING SET')
print(r2_score(y_test, pred_test_lasso_cv))

MSE: TRAINING SET
1.7383018712662366
SCORE: TRAINING SET
0.8957695259306389
MSE: TESTING SET
1.429917230190463
SCORE: TESTING SET
0.8899191365872575


In [36]:
print(lasso_cv.alpha_)

1.0


In [37]:
print(lasso_cv.intercept_)

3.036329724219719


In [38]:
print(lasso_cv.coef_)

[0.04678152 0.17490187 0.00142624]


# RidgeCV

In [39]:
from sklearn.linear_model import RidgeCV
regr_cv = RidgeCV(alphas=[0.01, 0.1, 1, 10.0])

regr_cv.fit(x_train, y_train) #RIDGE REGRESSION

pred_train_regr_cv= regr_cv.predict(x_train) #PREDICTIONS ON TRAINING SET

print('MSE: TRAINING SET')
print(np.sqrt(m.mean_squared_error(y_train,pred_train_regr_cv)))
print('SCORE: TRAINING SET')
print(r2_score(y_train, pred_train_regr_cv))

pred_test_regr_cv= regr_cv.predict(x_test) # PREDICTIONS ON TESTING TEST

print('MSE: TESTING SET')
print(np.sqrt(m.mean_squared_error(y_test,pred_test_regr_cv))) 
print('SCORE: TESTING SET')
print(r2_score(y_test, pred_test_regr_cv))


MSE: TRAINING SET
1.736902775492528
SCORE: TRAINING SET
0.8959372410233798
MSE: TESTING SET
1.411735322105487
SCORE: TESTING SET
0.8927007736376023


In [40]:
print(regr_cv.alpha_)

10.0


In [41]:
print(regr_cv.intercept_)

2.9087801619554288


In [42]:
print(regr_cv.coef_)

[0.04684319 0.17848636 0.00260306]
